lib/puppet/type.rb

```diff
@@ -2117,8 +2117,17 @@ def autorelation(rel_type, rel_catalog = nil)
     auto_rel = "eachauto#{rel_type}".to_sym
 
     self.class.send(auto_rel) { |type, block|
-      # Ignore any types we can't find, although that would be a bit odd.
-      next unless Puppet::Type.type(type)
+      if not Puppet::Type.type(type)
+        # We know it is not a type at this point, let's see if it exists as a defined type in the catalog
+        # We split and join on :: because defined types will likely have colon separators
+        # All defined types are of type :component
+        found_defines = rel_catalog.resources.select do |x|
+          x.type == :component && x.title_class == type.to_s.split('::').map(&:capitalize).join('::')
```

Comment:
@joshcooper There are methods named `builtin_type?()` and `builtin?()` (an alias for the former) on resource. Those differentiates between user defined types and "native" for the purpose of not having to evaluate the native resources (they have no code body that the compiler should evaluate). Note that app management adds new kinds of resources (predicate methods: `is_capability?`, `is_application_component?`) - notably "capabilities" (a.k.a "services") are implemented as native types.